### PR TITLE
Subway Graph

### PR DESCRIPTION
**Pull request by HadassahR**
- Added distance formula method + test
- Added method to calculate closest station to given coordinates + test

### CODE REVIEW

File: src/main/java/ClosestSubwayStations.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+public class ClosestSubwayStations {
+
+    public double calculateDistanceFormula (List<Double> locationCoords, List<Double> stationCoords) {
+        double calculateLatitude = Math.pow((stationCoords.get(0) - locationCoords.get(0)), 2);
+        double calculateLongitude = Math.pow((stationCoords.get(1) - locationCoords.get(1)), 2);
+        return Math.sqrt(calculateLatitude + calculateLongitude);
+    }
+
+    public List<SubwayStation.Station> calculateClosestStations (List<Double> originCoords, List<Double> destinationCoords, List<SubwayStation.Station> subwayStation) {
```
Comment (andrewoid) [inline src/main/java/ClosestSubwayStations.java:12]: Write a `getClosestStation(List<Double> coords)` method and just call it twice, it's simpler.

File: src/test/java/ClosestSubwayStationsTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        // when
+        List<SubwayStation.Station> closestStations = closestSubwayStations.calculateClosestStations(myLocation, myDestination, converter.getStationsList());
+
+        // then
+        Assert.assertEquals(expectedStations, closestStations);
+    }
+
+    @Test
+    public void invalid_getClosestSubwayStation() throws IOException {
```
Comment (andrewoid) [inline src/test/java/ClosestSubwayStationsTest.java:44]: You don't need to test a wrong answer...

File: src/main/java/CoordinateToStation.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    public SubwayStation.Station getClosestStation (List<Double> coordinates, List<SubwayStation.Station> subwayStation) {
+
+        double shortestDistanceFromCoords = Double.MAX_VALUE;
+        int indexCoords = 0;
+
+        for (SubwayStation.Station station : subwayStation) {
+            double distance = calculateDistanceFormula(coordinates, station.getGeometry().getCoordinates());
+            if (distance < shortestDistanceFromCoords){
+                shortestDistanceFromCoords = distance;
+                indexCoords = subwayStation.indexOf(station);
```
Comment (andrewoid) [inline src/main/java/CoordinateToStation.java:20]: Why are you saving the index of the Station when you can just save the Station?

File: src/main/java/SubwayNetwork.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        List<SubwayStation.Station> unvisitedStations = converter.getStationsList();
+        List<SubwayStation.Station> visitedStations = new ArrayList<>();
+
+        for (SubwayStation.Station station : unvisitedStations) {
+            station.getProperties().setVisited(false);
+            station.getProperties().setCurrent(false);
+            station.getProperties().setDistance(Integer.MAX_VALUE);
+        }
+
+        SubwayStation.Station currentStation = unvisitedStations.get(unvisitedStations.indexOf(origin));
```
Comment (andrewoid) [inline src/main/java/SubwayNetwork.java:30]: You have `origin`
You use origin to get the index in `unvisitedStations`
You then use the index to get the Station in `unvisitedStations`... which is `origin`

File: src/main/java/SubwayNetwork.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    private List<SubwayStation.Station> dijkstraAlgorithm (SubwayStation.Station origin, SubwayStation.Station destination) throws IOException {
+        Converter converter = new Converter();
+        List<SubwayStation.Station> unvisitedStations = converter.getStationsList();
+        List<SubwayStation.Station> visitedStations = new ArrayList<>();
+
+        for (SubwayStation.Station station : unvisitedStations) {
+            station.getProperties().setVisited(false);
+            station.getProperties().setCurrent(false);
```
Comment (andrewoid) [inline src/main/java/SubwayNetwork.java:26]: Why do you need to keep track of what the current station is with a boolean?

File: src/main/java/SubwayNetwork.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                }
+            }
+            stillPossibleConnections = smallestDistance != Integer.MAX_VALUE;
+        }
+        return visitedStations;
+    }
+
+    private List<SubwayStation.Station> tracePath (List <SubwayStation.Station> visitedStations, SubwayStation.Station origin, SubwayStation.Station destination) {
+        List<SubwayStation.Station> shortestPath = new ArrayList<>();
+        SubwayStation.Station destinationStation = visitedStations.get(visitedStations.indexOf(destination));
```
Comment (andrewoid) [inline src/main/java/SubwayNetwork.java:65]: Same as above

File: src/test/java/SubwayNetworkTest.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        Converter converter = new Converter();
+        SubwayStation.Station stationA = converter.getStations().get(264);
+        SubwayStation.Station stationB = converter.getStations().get(462);
+        SubwayNetwork subwayNetwork = new SubwayNetwork();
+
+        // when
+        List<SubwayStation.Station> result = subwayNetwork.findShortestPath(stationA, stationB);
+
+        // then
+        Assert.assertNotNull(result); // need expectedResult
```
Comment (andrewoid) [inline src/test/java/SubwayNetworkTest.java:37]: Add the expected result please.

File: src/main/java/SubwayNetwork.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    private Map<Integer, SubwayStation.Station> stations;
+
+    public List<SubwayStation.Station> findShortestPath(SubwayStation.Station origin, SubwayStation.Station destination) throws IOException {
+        List<SubwayStation.Station> visitedStations = dijkstraAlgorithm(origin, destination);
+        return tracePath(visitedStations, origin, destination);
+    }
+
+    private List<SubwayStation.Station> dijkstraAlgorithm (SubwayStation.Station origin, SubwayStation.Station destination) throws IOException {
+        Converter converter = new Converter();
```
Comment (andrewoid) [inline src/main/java/SubwayNetwork.java:20]: In order to get the Stations to pass into this method, you'd already had to have a Converter.

You are reading the files multiple times by doing this. You aren't allowed to do that.

File: src/test/java/SubwayNetworkTest.java (hunk 7 above)
Comment (HadassahR) [reply src/test/java/SubwayNetworkTest.java:37]: Working on this

**Comment by andrewoid** [comment]
```
/home/runner/work/SubwayProject/SubwayProject/src/main/java/SubwayNetwork.java:2: error: package jdk.internal.jline.internal does not exist
import jdk.internal.jline.internal.TestAccessible;
                                  ^
1 error

FAILURE: Build failed with an exception.
```

Get rid of that import.

**Pull request merged**